### DISCUSSION TITLE
I2C trouble

### DISCUSSION THREAD

**Discussion by DangarStu** [timmbogner/Farm-Data-Relay-System#237] [General]
I have a I2C Digital Wattmeter (https://www.dfrobot.com/product-1827.html) connected to a Sailor Hat ESP32 and I noticed that it stopped initialising after I made some software changes. I later tracked it down to having turned OLED off. When I turned that on again it started working again. I'll do some futher investigation but just wanted to ask if this is a known issue that I need to handle in a specific way.

Stu

**Comment by timmbogner** [comment]
Hey Stu, 
I don't have a setup for testing available this week, but I think I have a best-guess for your issue. [Here](https://github.com/search?q=repo%3Atimmbogner%2FFarm-Data-Relay-System%20use_i2c&type=code) you can see the code that enables I2C to work within FDRS. If you aren't using I2C within FDRS, you'll need to include `Wire.h` and call `Wire.begin()` on your own. If that's not the problem, post your sketch here and that should shine more light on the problem.

I do apologize for my lack of responsiveness. I've just had too much on my plate lately to really focus on the project. Hopefully I'll have more freedom this winter! I really appreciate your interest in the project and the feedback you had about configurations etc. 
Thanks! 
Timm

**Reply by aviateur17** [reply]
I've been busy as well setting up my whole home battery with inverter and solar panels to shift all my grid usage to off peak and save some money using off peak rates.  Instead of using a Chinese web server to see all the data I'm using a PLC to read data via modbus from the inverter and send it to MQTT and then display the data via Home Assistant.   There's always more to do but I'm getting close to being at a place where I'm comfortable.   All that to say I hope that I can spend more time in this project soon.  I'll look at that I2C issue probably tomorrow.
![PXL_20241006_173206197](https://github.com/user-attachments/assets/4d16fbf4-273d-44e4-a043-90bac68d2afc)
![PXL_20241006_173029466](https://github.com/user-attachments/assets/3f672670-8143-4434-8f0a-75181a6b3f72)

**Reply by DangarStu** [reply]
Thanks Timm

Thanks for your reply. Certainly don't every feel any need to apologise, everyone is busy and you've already done so much making FDRS! 

Here is the beginning of the sketch where I include `Wire.h` and call `Wire.begin()` but get the `INA219 begin failed` error. In that code you can see I pasted some boiler plate code to scan for all devices in case it was an address problem but when I don't have `USE_OLED` defined it finds nothing and when I do it finds both devices.

```
#include "fdrs_node_config.h"
#include <fdrs_node.h>
#include <Wire.h>
#include "DFRobot_INA219.h"

#define FLOAT_SWITCH_PIN  35

DFRobot_INA219_IIC     ina219(&Wire, INA219_I2C_ADDRESS4);

void setup() {
  pinMode(FLOAT_SWITCH_PIN, INPUT_PULLDOWN);
  analogReadResolution(12); // Set ADC resolution to 12 bits
  beginFDRS();
   byte count = 0;
      for (byte address = 8; address < 120; address++) {
        Wire.beginTransmission(address);
        if (Wire.endTransmission() == 0) {
            DBG("I2C device found at address 0x");
            DBG(address);
            count++;
        }
    }

    if (count == 0) {
        DBG("No I2C devices found.");
    } else {
        DBG("I2C scan complete.");
    }

  //Initialize the voltage and current sensor
  while(ina219.begin() != true) {
      DBG("INA219 begin failed");
      delay(2000);
  }
}
```

Stu

**Reply by DangarStu** [reply]
> I've been busy as well setting up my whole home battery with inverter and solar panels to shift all my grid usage to off peak and save some money using off peak rates. Instead of using a Chinese web server to see all the data I'm using a PLC to read data via modbus from the inverter and send it to MQTT and then display the data via Home Assistant. There's always more to do but I'm getting close to being at a place where I'm comfortable. All that to say I hope that I can spend more time in this project soon. I'll look at that I2C issue probably tomorrow. ![PXL_20241006_173206197](https://private-user-images.githubusercontent.com/61305397/390594007-4d16fbf4-273d-44e4-a043-90bac68d2afc.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI3ODkxODYsIm5iZiI6MTczMjc4ODg4NiwicGF0aCI6Ii82MTMwNTM5Ny8zOTA1OTQwMDctNGQxNmZiZjQtMjczZC00NGU0LWEwNDMtOTBiYWM2OGQyYWZjLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI4VDEwMTQ0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRhMDg0Zjc4N2E3Njk2OWE2ZjA5MWIxMGE2MWNmZjRjMWU5MGQ2YzQ2OTAzZWM3N2QwYzM0MzA5M2NmMTE0ZTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XerHOE17-h1K2ypX8PzVOC2N7E11WsgUZyHGamfeNwg) ![PXL_20241006_173029466](https://private-user-images.githubusercontent.com/61305397/390594014-3f672670-8143-4434-8f0a-75181a6b3f72.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI3ODkxODYsIm5iZiI6MTczMjc4ODg4NiwicGF0aCI6Ii82MTMwNTM5Ny8zOTA1OTQwMTQtM2Y2NzI2NzAtODE0My00NDM0LThmMGEtNzUxODFhNmIzZjcyLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI4VDEwMTQ0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExZDRiZDdiY2Q0MmEzMjMzOTg5ODM5MWZlMjUwZTE1ZmI0ZDM3MTg0ODdhM2YxYjQwNDNlM2MyMjMwMmMzMDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ikFfhc14HVGs-UM9EuCd0Qm-Gr-rZ-_1TmZkBCAFnOQ)

That's quite the solar setup! You've reminded me I still need to fix my inverter...

**Reply by aviateur17** [reply]
In your fdrs_node_config.h file you can add 
```
#define USE_I2C
```
or you can insert that same #define after your #include of fdrs_node_config.h line.  Make sure you define it before the line that says #include <fdrs_node.h>.   Then in your code you can comment out or remove your #define <wire.h> as that is done in the fdrs_node.h file.   Make sure you define your I2C pins in the fdrs_node_config.h file.
```
// I2C - OLED or RTC
#define I2C_SDA 5
#define I2C_SCL 6
```
Replace 5 and 6 with whatever pins you are using on your board.

**Reply by DangarStu** [reply]
> In your fdrs_node_config.h file you can add
> 
> ```
> #define USE_I2C
> ```
> 
> or you can insert that same #define after your #include of fdrs_node_config.h line. Make sure you define it before the line that says #include <fdrs_node.h>. Then in your code you can comment out or remove your #define <wire.h> as that is done in the fdrs_node.h file. Make sure you define your I2C pins in the fdrs_node_config.h file.
> 
> ```
> // I2C - OLED or RTC
> #define I2C_SDA 5
> #define I2C_SCL 6
> ```
> 
> Replace 5 and 6 with whatever pins you are using on your board.

Thank you very much! I will experiment some more tomorrow and report back.

**Reply by DangarStu** [reply]
> #define USE_I2C

That worked perfectly. It looks like I was missing the setting of the SDA and SCL pins without the define:

```
#ifdef USE_I2C
  Wire.begin(I2C_SDA, I2C_SCL);
#endif
```

It's great to understand who it all hangs together so I can avoid those duplicate includes etc.

Thanks again,
Stu

**Reply by timmbogner** [reply]
@aviateur17 Awesome solar setup! I have been stockpiling parts and batteries to build a network of small solar lamps that can synchronized via FDRS... a tiny version of your project!

@DangarStu Glad you got it worked out. My Sailor Hat arrived last week... I appreciate how they packaged it in paper instead of the usual silver plastic, but hat's about as far as I've gotten. Definitely post some updates on your project if you feel inclined!
